Decoy Implementation and Additional Selection Methods in Snekmer Learn/Apply

jjacobson95 opened the pull request:
This PR brings Snekmer Learn/Apply up to date with numerous developments that have occurred since the last update.

These updates include:

- Reverse Sequence Decoys
- Family Specific Thresholds 
- Additional Selection methods (such as 'greatest_distance')
- Added an output directory from Learn called "apply_inputs" that consolidates the 3 key files needed for Apply.
- Additional misc bug fixes.


To do: use `snakefmt` for linting - I did a quick test and it broke formatting, so I'll come back to this when we want to merge the PR.


---
**Additional possible updates in this or a future PR include**:
- Many to Many 'Apply': allow users to use sequences for hits instead of families.
- Inclusion of values below the decoy threshold. Currently these are excluded, we should instead add a flag and let the user decide if they want to use them.
- Multi-hit function: Allow there to be multiple hits for a query sequence. This may be done by utilizing the family specific cutoffs and allowing all scores above a threshold to pass.

christinehc committed:
build: create pyproject.toml for pypi install (fixes #119)

changelog:
- created pyproject.toml for snekmer installation
- remove setup.py (no longer needed / deprecated)
- `pip install snekmer` should now work`

christinehc committed:
build: update pip install links for snekmer to pypi

christinehc committed:
build: require python 3.9+

jjacobson95 committed:
Pre-generated Decoy input integrated. Summary stats and plots integrated. Old confidence method disabled. Next steps: Work on family based threshold confidence method

jjacobson95 committed:
code for all methods in this commit. Still working on reintegrating confidence

jjacobson95 committed:
Saving changes before big update

jjacobson95 committed:
Major updates. Multiple methods introduced. Confidence method reinstated. More work to be done

jjacobson95 committed:
Cleaning up some unused code in Learn

jjacobson95 committed:
previous version while in works

jjacobson95 committed:
Learn with reverse seqs as decoys. Tons of commented code and junk

jjacobson95 committed:
Removed redundant code and comments

jjacobson95 committed:
About to do more code clean up. Just storing as is in case I break anything

jjacobson95 committed:
All changes to get Snekmer Learn Apply up to date. This has reverse sequence decoys, family specific thresholds, and multiple selection methods

jjacobson95 committed:
Restructured so snakefmt would work

jjacobson95 committed:
Merge branch 'main' into learn_decoy_implementation

tnitka reviewed (changes requested):
We definitely need to uptick the minor version number here. I'm not sure why CI is failing on this branch, so we should probably investigate that too if you haven't already.

jjacobson95 commented:
Good call on the version uptick.  I'm currently investigating why its failing on CI, I think its due to me using a snakefmt version compatible with python3.10 - I looked on the super-linter page and they use a newer version of python, so I'm working on this.

jjacobson95 committed:
Testing snakefmt

jjacobson95 committed:
Formatted with correct version of snakefmt and black

jjacobson95 commented:
Okay, I have no idea why its failing now. Snakefmt and black are the correct versions and it passes locally.  At this point I think the best option is to move the long code chunks to modular scripts and call them from snakemake. @christinehc recommended this fix earlier and I agree its a cleaner and better long term option.

jjacobson95 committed:
All Large code chunks moved to modular scripts

jjacobson95 committed:
Version uptick

jjacobson95 committed:
Linted.

jjacobson95 committed:
Pinning the env to this branch for tests

tnitka committed:
Update docs to reflect Learn/Apply decoy scoring

jjacobson95 commented:
Tests passed! Before merging, we should unpin the test environment so its not linked to a branch. This will cause a test failure but we've now manually confirmed that it passes.

@tnitka please be sure to include all of the new config variables that are required in the docs update. I also added an output to `Learn`, this is called `apply_inputs` which contains the following subdirectories with the corresponding files: confidence, stats, counts. This was to reduce some intermediate steps between learn and apply (and eventually could be used in a "quick start" snekmer learn apply function).

Screenshot of outputs:
---
<img width="1219" alt="Screenshot 2025-05-22 at 12 57 08 PM" src="https://github.com/user-attachments/assets/c7e615f8-6df8-4ec5-ac92-5b8c6d253555" />

Config Screenshot for learnapp params:
---
<img width="669" alt="Screenshot 2025-05-22 at 1 03 34 PM" src="https://github.com/user-attachments/assets/a26c2090-11c8-4a36-8af2-4b8885060e04" />

tnitka commented:
I think I have the rest of the updates to docs finished, but the tutorial appears to expect separate input directories for learn and apply even though the repository has only one input directory for the tutorial? I can split them if this is how the tutorial is meant to work (currently I can't regenerate the outputs to include in the docs, as the notebook isn't finding the inputs).

jjacobson95 commented:
Yeah, the tutorial is incredibly out of date at this point. Feel free to change anything you think is necessary.

The CI test just checks if this can run so only has 1 set of inputs (learn and apply on the same sequences). We could choose to do the tutorial this way just for demonstration purposes, but I think it might be better to have a separate input dir for apply

tnitka committed:
Update Learn/Apply tutorial for confidence-only settings

tnitka commented:
The tutorial works now (currently uses 8 proteomes for learn and 4 for apply), but still needs to be updated to actually use the new family thresholds.

tnitka committed:
Update Learn/Apply tutorial for decoy thresholds.

jjacobson95 commented:
Just a note, the docker container has been built and uploaded to [DockerHub](https://hub.docker.com/r/jjacobson95/snekmer). Tara has tested this on her end as well and is updating the docs for this.

This docker container is much more user friendly than the original. Originally a user would have to `exec` into a container and use it like a terminal with manual file handling. Now with the latest implementation a user can download it and use it just like you would use snekmer on the local command line.

Instructions:
---
Setup directory structure like you normally would for snekmer, then:
**Download docker image from DockerHub**
```
docker pull jjacobson95/snekmer:latest
```
**Run Any Snekmer Mode**
```
docker run --rm -v "$(pwd)":/data -w /data jjacobson95/snekmer:latest {mode} {args}
```
**Example: Snekmer Learn with 8 cores**
```
docker run --rm -v "$(pwd)":/data -w /data jjacobson95/snekmer:latest learn --cores 8
```

tnitka committed:
Update docs for current version of Docker image.

tnitka committed:
Change tutorial data so Snekmer has enough training data without requiring excessive time to run

tnitka reviewed (approved):
I don't see anything else that we need to address right now, unless Jason or Christine sees an issue this should be ready to merge.

biodataganache commented:
I'm looking through the PR and branch right now and see some minor issues with the documentation (including the learn/apply Jupyter notebook).

For the Jupyter notebook (snekmer_learnapp_tutorial): 
The Jupyter notebook is not really a beginner's tutorial. It's for people who want to understand what snekmer is doing at each step in some depth. I think we should probably specify this in the documentation.
TIGRFAMs.ann should be renamed to something more generic. 
Formats for the input files should be specified and/or linked (how is the .ann file to be formatted?)
More motivation for the input needs to be given - what does the pipeline do? Why are each of the files necessary?

Learn and apply have no report html output. For apply this seems like it could be especially helpful and actually necessary since it's not clear how to access results. The 'accessing results' section of the documentation is insufficient since it doesn't give any idea about what these results are supposed to be (and why do we want to make the user chase down their results?)

We have some pre-calculated learn matrices. I don't think it makes sense to distribute them in the repo - but I do think it makes sense to give users access to them. I think that we should reformulate the documentation/demos for apply so that there's a simple 'here's how you apply TIGRFAMs learned matrix to your dataset' - this should include the html report

While I'm on it I don't love the way there's no single output to Snekmer. I would like in a near future version to have one summary file that gets written to the same directory you start in (as default behavior) that way we don't make the users go on a hunt for their results.

I may have other issues as well

christinehc commented on snekmer/rules/apply.smk:
```diff
@@ -49,6 +49,8 @@ from scipy import spatial
 from scipy.stats import rankdata
 
 import snekmer as skm
+from pkg_resources import resource_filename
```
Note: `pkg_resources` is deprecated and removed in Python 3.12. Use `importlib` instead. See: https://setuptools.pypa.io/en/latest/pkg_resources.html

christinehc commented on snekmer/scripts/learn_eval_apply_reverse_seqs.py:
```diff
@@ -0,0 +1,269 @@
+# Imports
+# ---------------------------------------------------------
+
+import pickle
```
Style comment: imports can be sorted using something like isort

christinehc replied:
Also, these seem to be a lot of imports, and about the same imports in every file. Are they all being used? Please remove any that are not

jjacobson95 replied:
Done!

christinehc commented on snekmer/scripts/learn_eval_apply_reverse_seqs.py:
```diff
@@ -0,0 +1,269 @@
+
+
+class KmerCompare:
+    """
```
Docstrings are not in a standard docstring format, especially the indentation. Please refer to: https://peps.python.org/pep-0257/

christinehc replied:
Please apply this comment to all docstrings in this file

jjacobson95 replied:
Ah yes, this was due to snakefmt reshaping everything. I'll standardize them now that they aren't being linted by snakefmt anymore.

christinehc commented on snekmer/scripts/learn_eval_apply_reverse_seqs.py:
```diff
@@ -0,0 +1,269 @@
+        for i, seqid in enumerate(seqs):
+            self.seqAnnot[seqid] = anns[i]
+
+    def generate_reverse_kmerCounts(self):
```
Please be consistent about naming; most of the attributes here seem to be camelCase but this is snake_case + camelCase. Consistency is preferable.

Note that the rest of Snekmer is snake_case so I'd prefer adherence to that convention, but if nothing else, self-consistency within one python file is the minimum

christinehc commented on snekmer/scripts/learn_eval_apply_reverse_seqs.py:
```diff
@@ -0,0 +1,269 @@
+"""
+        kmerList, df = skm.io.load_npz(self.inputData)
+        self.kmerList = kmerList[0]
+        seqids = df["sequence_id"]
```
Again re:variable naming consistency. seq_ids or seqIds, not seqids

christinehc commented on snekmer/scripts/learn_eval_apply_reverse_seqs.py:
```diff
@@ -0,0 +1,269 @@
+"""
+
+    def __init__(
+        self, compareAssociations, annotationFiles, inputData, outputPath
```
Since types are already specified above: I'd encourage [type hinting](https://docs.python.org/3/library/typing.html) in all function definitions.

christinehc commented on snekmer/scripts/learn_eval_apply_reverse_seqs.py:
General comment: I would suggest moving all object and function definitions to an actual package module file and importing them using `from snekmer.apply import <...>`

christinehc commented on snekmer/scripts/learn_eval_apply_reverse_seqs.py:
```diff
@@ -0,0 +1,269 @@
+Returns:
+    DataFrame: Formatted kmer counts DataFrame.
+"""
+        if len(str(kmerCounts.columns.values[10])) == len(
```
Could be simplified to: `compareCheck = len(str(kmerCounts.columns.values[10])) == len(str(self.kmerCountTotals.columns.values[10]))`

christinehc commented on snekmer/scripts/learn_eval_apply_sequences.py:
```diff
@@ -0,0 +1,332 @@
+# ---------------------------------------------------------
+# Imports
+# ---------------------------------------------------------
+
```
See previous comments re:import and docstring formatting

christinehc commented on snekmer/scripts/learn_eval_apply_sequences.py:
```diff
@@ -0,0 +1,332 @@
+# ---------------------------------------------------------
+
+
+class KmerCompare:
```
Is this not the same object defined in the previous file? Hence why I suggest putting this into a Snekmer module and importing into these files instead of redefining twice

If there are differences between this KmerCompare and the previous one, are there ways to programmatically enable these two slightly different implementations in a singular object that can, again, be imported from the snekmer package?

jjacobson95 replied:
These two should not have had the same name. There were some similarities but some pretty major differences too. Renamed as they do different things

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+# ---------------------------------------------------------
```
See previous file re:import sorting and docstring formatting

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+# Run script
+# ---------------------------------------------------------
+
+class Evaluator:
```
See previous comment re:moving objects and functions to package modules for import

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+
+        # Load thresholds
+
+        thresholds_df = pd.read_csv(
```
Again, please stick to one convention for variable names (snake_case vs. camelCase)

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+        )
+
+        # Load thresholds
+
```
Remove extraneous line break

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+
+            self.threshold_dict = threshold_dict  # Store for later use
+
+            # Apply the selection method
```
Why is the indentation off for this comment?

jjacobson95 replied:
snakefmt previously shifted comments in an unpredictable way. Fixed all inconsistencies that I found though.

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+                tf.append("T")
+            else:
+                tf.append("F")
+                # Determine if the sequence is known or unknown
```
There are several more instances of inconsistent indentation with the comments. Please go through and fix all of them. (Note: the one above needs to be flush with the line of code below that it is referring to)

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+        # Filter out rows where Prediction is None
+        valid_predictions = diffDataframe.dropna(subset=["Prediction"])
+
+        known_true_diffDataframe = valid_predictions[
```
I don't want to point out all instances of variable inconsistency, but to me, this one is an especially egregious example. Please choose one convention. Additionally, variable names should generally not be so long and unwieldy. This can easily be compressed to something like `known_true_diff_df`, `known_true_diffs`, etc.

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+        updatedData.fillna(0, inplace=True)
+        return updatedData
+
+    def saveResults(
```
Does this really warrant its own separate function?

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+
+        elif selectionMethod == "top_hit" and threshold_type is not None:
+
+            def get_topTwo(row):
```
This seems to have a lot (at least the formulation of the series at the end) in common with the previous `get_topTwo` function. I would streamline these so you're not duplicating so much code

tnitka committed:
Add report output to Apply

tnitka committed:
Linted

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,709 @@
+
+        return diffDataframe
+
+    def apply_selectionMethod(
```
This function is very long (~250 lines) with multiple if/elif conditionals and not much commentary on what each stream is doing. This needs to be broken down into multiple functions and better documented to describe steps

christinehc commented on snekmer/scripts/learn_fragment.py:
```diff
@@ -0,0 +1,110 @@
+# ---------------------------------------------------------
+# Imports
+# ---------------------------------------------------------
+
```
See previous comments about imports

christinehc commented on snekmer/scripts/learn_fragment.py:
```diff
@@ -0,0 +1,110 @@
+        for i in range(0, len(sequence) - actual_length + 1):
+            frags.append(sequence[i : i + actual_length])
+
+            # Filter fragments based on min_length
```
Please ensure comments are at the same indentation level as the code to which they refer

christinehc commented on snekmer/scripts/learn_fragment.py:
```diff
@@ -0,0 +1,110 @@
+
+random.seed(params.seed)  # setting the seed for randomness
+
+def fragment(sequence, version, length, location, min_length):
```
Please move this to a module (e.g. learn.py)

christinehc commented on snekmer/scripts/learn_learn.py:
```diff
@@ -0,0 +1,235 @@
+# ---------------------------------------------------------
+# Imports
+# ---------------------------------------------------------
+
```
See previous commentary about imports and docstring formatting.

christinehc commented on snekmer/scripts/learn_learn.py:
```diff
@@ -0,0 +1,235 @@
+It creates a kmer-count matrix for each input file.
+
+Attributes:
+annotation (list): A list to store annotations loaded from files.
```
See previous comment about type hinting (since the types are annotated in the docstrings anyway)

christinehc commented on snekmer/scripts/learn_learn.py:
```diff
@@ -0,0 +1,235 @@
+Args:
+    input_annotation (list): List of file paths containing annotations.
+"""
+        for f in input_annotation:
```
Please see previous comments about consistency in variable naming and apply to this entire file

christinehc commented on snekmer/scripts/learn_learn.py:
```diff
@@ -0,0 +1,235 @@
+# Run script
+# ---------------------------------------------------------
+
+class Library:
```
See previous comment about moving objects to snekmer modules

christinehc commented on snekmer/scripts/learn_merge.py:
```diff
@@ -0,0 +1,206 @@
+# ---------------------------------------------------------
+# Imports
+# ---------------------------------------------------------
+
```
See previous comments about imports and docstring formatting

christinehc commented on snekmer/scripts/learn_reverse_decoy_evaluations.py:
```diff
@@ -0,0 +1,242 @@
+# ---------------------------------------------------------
+# Imports
+# ---------------------------------------------------------
+
```
See previous comments about imports

christinehc commented on snekmer/scripts/learn_reverse_decoy_evaluations.py:
```diff
@@ -0,0 +1,242 @@
+
+
+def save_stats_to_csv(combinedStats, csv_file):
+
```
Please add docstrings and type hinting to every function

christinehc commented on snekmer/scripts/learn_reverse_decoy_evaluations.py:
```diff
@@ -0,0 +1,242 @@
+# ---------------------------------------------------------
+
+
+def load_existing_stats_from_csv(csv_file):
```
See previous comments about moving functions to package modules

christinehc commented on snekmer/scripts/learn_reverse_decoy_evaluations.py:
```diff
@@ -0,0 +1,242 @@
+    return combinedStats
+
+
+def save_stats_to_csv(combinedStats, csv_file):
```
See previous comments about consistency in variable and function naming

christinehc commented on snekmer/scripts/learn_reverse_decoy_evaluations.py:
```diff
@@ -0,0 +1,242 @@
+
+
+def generateFamilyStatistics(combinedStats):
+    statsData = {
```
How much of these are necessary? Can we do min, 25, 50, 75, max?

jjacobson95 replied:
These are used as the values for family specific thresholds, so I think we want to keep most of them?  I removed a few of the extra percentile values.

christinehc reviewed (changes requested):
I've left some comments re:code formatting, style, and some refactoring that needs to be completed before we can finalize the package. Let me know if you have any questions.

tnitka committed:
Update Learn/Apply tutorial text

jjacobson95 committed:
Changed to CamelCase across all files. Reduced imports

jjacobson95 committed:
Merge remote-tracking branch 'refs/remotes/origin/learn_decoy_implementation' into learn_decoy_implementation

jjacobson95 committed:
Docstrings and Typing Added. More updates in progress

jjacobson95 committed:
Whooo that was a lot... Migrated similar code chunks to a module called learn.py

jjacobson95 committed:
Renamed all vars. Migrated more duplicate code to modular scripts. Added final docstrings

jjacobson95 commented:
Just an update, I think I've resolved the majority of @christinehc's concerns. Overall, the scripts now have a lot more consistency and should be up to date with modern python conventions. Please let me know you think I missed anything, there was a substantial amount to update.

Still to do:
- ~~Update `package_resources` to `importlib`~~
- ~~Create a single file output to snekmer apply. This will essentially just be concatenating all of the kmer-summary files in the 'output/apply' directory~~

@biodataganache what do you think about a `quick-la` option ("la" stands for learn/apply). Is this worth time to develop? If so, should it be part of this PR?

This would require the 3 inputs shown below and the config file would be pulled from the package and hidden from the user.

`snekmer quick-la --annotation {annotation_file} --learn {learn_dir or learn.fasta} --apply {apply_dir or apply.fasta}`

jjacobson95 committed:
Migrated to importlib

jjacobson95 committed:
Added single output for Apply in base dir. Added outputs for learn called 'apply_inputs' in base dir

jjacobson95 committed:
Removed unused variables. fixed test bug

tnitka committed:
Update Learn/Apply tutorial

jjacobson95 committed:
updated resources

jjacobson95 committed:
update

jjacobson95 committed:
Docs update for latest params

christinehc committed:
Merge main changes to package (pypi) (#131)

* Implement bad debugging practive to be reverted in next commit

* Fix input handling and remove bad debugging practice

* Fix compressed input handling in motif script

* Fix conflict between dataframe and ndarray usage in motif_motif.py

* Use common basis for all proteins in motif

* Fix input error in motif snakefile

* Fix input error in motif snakefile

* Fix input error in motif script

* Fix input processing error in motif

* Fix input processing error in motif

* Fix input processing error in motif

* Fix scoring error in motif script

* Fix error in motif scoring

* Fix error in motif scoring

* Fix error in motif scoring

* Fix type error in motif script

* Fix matrix shape error)

* Fix matrix shape error in motif script

* Fix input handling error in motif script

* Fix input handling error in motif script

* Fix input handling error in motif script

* Correct input matrix column used for family labels

* Correct input matrix column used for family labels

* Fix input handling error in motif script

* Fix input handling error in motif script

* Fix input handling error in motif script

* Fix class error in motif script

* Fix class error in motif script

* Fix class error in motif script

* Fix class error in motif script

* Fix issue with DataFrame not being converted to ndarray

* Fix issue with DataFrame  being converted to empty ndarray

* Fix typographical error in motif script that prevented code execution

* Correct error in selecting data labels

* Fix type error in scoring parameters

* Fix type error in scoring parameters

* Fix type error in scoring parameters

* Fix type error in scoring parameters

* Fix type error in scoring parameters

* Fix type error in motif.py

* Fix error passing params to scorer

* Fix error passing params to scorer

* Fix error passing params to scorer

* Fix error passing params to scorer

* Fix error passing params to scorer

* Fix error passing params to scorer

* Fix error passing params to scorer

* Fix error in motif object

* Fix valueerror in motif script

* Change motif object to use pandas dataframe instead of numpy ndarray

* Fix formatting errors in motif

* Fix formatting errors in motif

* Fix formatting errors in motif

* Fix input error in motif script

* Fix input error in motif script

* Fix input error in motif script

* Fix error passing params to motif module

* Fix error passing params to score module

* Fix error passing params to score module

* Fix error fetching output from score module

* Fix error fetching output from score module

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix error tabulating permutation scores

* Fix reference error in motif module

* Fix reference error in motif module

* Fix array dimension error in motif script

* Fix array dimension error in motif script

* Fix array dimension error in motif script

* Fix array dimension error in motif script

* Fix array dimension error in motif script

* Allow array concatenation in motif script

* Fix array dimension error in motif script

* Fix array dimension error in motif script

* Fix array dimension error in motif script

* Fix array dimension error in motif script

* Change p_values class in motif module to use dataframes

* Fix indexerror in motif module

* Fix type error in motif module

* Fix type error in motif module

* Fix type error in motif module

* Fix type error in motif module

* Fix type error in motif module

* Fix error uncompressing input

* Fix error uncompressing input

* Fix error uncompressing input

* Fix error uncompressing input

* Fix error uncompressing input

* Fix error in p_values class

* Fix input data typing error

* Fix typing error in motif module

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Fix type error in motif

* Remove redundant output

* Prevent snakemake from expecting redundant output

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error that causes empty output file

* Fix error preventing kmers from being included in output

* Fix error preventing kmers from being included in output

* Fix error preventing all kmers from being scored

* Fix indexing errors

* Fix indexing errors

* Fix indexing errors

* Fix indexing errors

* Fix indexing errors

* Fix indexing errors

* Fix indexing errors

* Fix indexing errors

* Fix indexing errors

* Fix indexing errors

* Explicitly name columns to fix KeyError in motif

* Explicitly name columns to fix KeyError in motif

* Explicitly name columns to fix KeyError in motif

* Explicitly name columns to fix KeyError in motif

* Fix issue selecting kmer sequence in motif

* Fix issue selecting kmer sequence in motif

* Change iteration from DataFrame to NDArray

* Change iteration from DataFrame to NDArray

* Change iteration from DataFrame to NDArray

* Change iteration from DataFrame to NDArray

* Change iteration from DataFrame to NDArray

* Change iteration from DataFrame to NDArray

* Fix error fetching kmer scores from score

* Fix error fetching kmer scores from score

* Fix error fetching kmer scores from score

* Fix type error in p value calculation

* Remove redundant code

* Fix issue in motif rule that was sometimes causing MissingInputException

* Ensure that all permutation scores are compared to the real score when running snekmer motif

* Ensure that all permutation scores are compared to the real score when running snekmer motif

* Ensure that all permutation scores are compared to the real score when running snekmer motif

* Add scores as output from motif

* Remove score output from motif due to I/O issues

* Remove score output from motif due to I/O issues

* Temporarily print permutation scores to check whether they are identical

* Remove printing as it is no longer necessary

* Remove unnecessary conversions in motif script

* Fix syntax in motif script

* Add scores as output from motif

* Fix issue causing scores to be the same across iterations in motif

* Remove unused code

* Change motif output to be sorted by p value

* Fix motif output sorting order to put most significant kmers first

* Fix issue parsing kmers when called with k=2

* Fix motif output sorting order to put most significant kmers first

* Change output sorting to use p value first followed by score on real data

* Fix motif output sorting order to put highest scoring kmers first for a given p value

* Fix formatting of csv containing scores from motif iterations

* Fix issue causing motif to read kmer NA as np.nan

* Expand definition of false positives in motif to scores greater than or equal to the real score

* Fix issue causing too few scoring iterations to be compared in motif

* Remove minimum family size from motif

* Remove unnecessary family size check from motif snakefile

* Fix issue sorting negative scores in motif

* Fix issue sorting kmers with negative weights in motif output

* Parallelize the motif workflow

* Fix error sorting output of motif workflow

* Remove redundant code from motif workflow to slightly reduce memory usage and computational cost

* Reduce memory usage in motif and model

* Further reduce memory usage in model

* Fix an issue affecting rescoring results

* Reduce memory usage in motif

* Reduce memory usage of motif module

* Reduce peak memory usage during motif

* Fix error normalizing kmer weights in motif

* Fix error normalizing kmer weights in motif

* Fix error normalizing kmer weights in motif

* Fix error normalizing kmer weights in motif

* Fix error normalizing scores in motif

* Add preselection step to motif workflow

* Fix score scaling error in motif

* Fix error in motif preselection step

* Fix error calculating p-values in motif

* Apply recursive feature elimination during motif preselection

* Adjust motif preselection stopping criterion to improve results

* Decrease RFE step size in motif preselection step

* Remove redundant code

* fixup! Remove redundant code

* Format/lint code

* Format/lint motif snakefile

* Fix whitespace error in snakefmt output

* Format/lint snakefile

* update action.yml

* Update test and fix command line parser error introduced during rebase

* fixup! Update test and fix command line parser error introduced during rebase

* Add motif test to CI workflow

* Correct snekmer motif test environment

* chore: update _version.py

* Add Motif tutorial

* Update docs to include motif

* Add model from RFE as output in motif

* Update documentation for motif

* Fix formatting

* Add motif report output

* docs: add demo pages for learn/apply and motif

changelog:
- RTD pages now include notebook-formatted pages with the learn/apply and motif demos.
- Formatted notebooks for compatibility with sphinx notebook formatting.
- Cleaned up notebooks, e.g. removing extraneous import statements

* docs: clean up files and update/create symlinks

changelog:
- removed ~HEAD notebook file
- moved learn/apply and motif tutorial notebooks to docs folders to add them into sphinx documentation
- created symlinks for the above notebooks so that notebooks are still accessible from the resources directory

* Fix formatting

* Remove redundant code from motif result script

* move Motif tutorial into separate directory with more informative config parameters

* Add motif to README.md

* Updated CI runner. Updated Learn/Apply Test

* Changing pin from motif to learn, will need to remove pin entirely soon

* Working on CI action workflow

* Debuging Build env for CI

* Still debugging

* Still debugging

* Still debugging

* Still debugging

* Old method not working. Testing manual build

* This 'should' resovle the build env

* Build is working, now updating dependancies

* Fixing annotation file for test

* Tests passed when pinned to @learn_decoy_implementation branch. Setting to main so this won't be branch specific.

---------

Co-authored-by: Tara Nitka <tara.nitka@pnnl.gov>
Co-authored-by: Tara A. Nitka <86532598+tnitka@users.noreply.github.com>
Co-authored-by: Jeremy <jeremy.jacobson3402@gmail.com>

christinehc committed:
build: update build requirements

changelog:
- build: remove constraints on numpy and pandas -- the fixed versions create package dependency conflicts. has not been tested against learn/apply workflow to see if issues arise
- build: require python<3.12 (`asyncore` is deprecated in 3.12, which raises an error when running snekmer)

christinehc committed:
chore: resolve merge conflicts

christinehc commented on .test/test-env.yml:
Note: needs to be pointing to this branch for the purposes of the PR, but should be reverted to main once the merge happens

tnitka committed:
Resolve Learn/Apply tutorial merge conflicts

christinehc committed:
docs: update install instructions with pypi package

christinehc commented on .test/config_learnapp.yaml:
Meta question both here and in the example files (./resources), do we need a separate learn/apply config? Can we just include the learn/apply parameters in one config.yaml and consolidate?

jjacobson95 replied:
In our test, we do require separate config files. The input directories are defined in the config file and our cluster/model/search tests require a different input directory to be defined than learn/apply.

I did update the regular config to include the learn and apply parameters as well

christinehc committed:
docs: add minor formatting tweaks for readability

christinehc commented on resources/LA_config.yaml:
Again, does this need to be a separate config file? Can we consolidate the provided config files?

christinehc replied:
And see my note on the other LA_config re: "learnapp"

jjacobson95 replied:
Consolidated the three config files in the resources into one.

christinehc commented on docs/source/tutorial/snekmer_learnapp_tutorial.ipynb:
Would like to see included an explanation of the annotation file in the beginning portion

christinehc commented on .test/test-env.yml:
Minor note that we should remember to change this right before we merge so that it's pointing to either the main branch or pypi

christinehc commented on docs/source/index.rst:
Can we adhere to lowercase/ideally snake_case for column names, so "family" ?

christinehc commented on .test/config_learnapp.yaml:
Another comment-- I would prefer if we either (a) separated learn and apply parameters to a "learn" and "apply" section in the config, or else (b) consolidated all params into "learn_apply" (vs. "learnapp"). I realize the impulse for name shortening in the codebase, but for usability (esp. since the user directly interacts with the config), I dislike the truncation.

jjacobson95 replied:
We could separate them, however there would then be duplicate values/parameters. Learn/Apply is a paired pipeline and some of the parameters in Learn must also be used in Apply. For example, if selection method is set differently in Learn and Apply, there would not be an error message, but the results would be highly inaccurate.

I'll rename the section to learn_apply

christinehc commented on snekmer/scripts/learn_evaluate_sequences.py:
```diff
@@ -0,0 +1,476 @@
+        ratio_total_dist = true_total_dist / (true_total_dist + false_total_dist)
+        raw_total_sum = (true_total_dist + false_total_dist).copy()
+        new_index = pd.Index(
+            [round(i, 2) for i in pd.np.arange(0, 1.01, 0.01)],
```
Should be `np.arange`?

christinehc replied:
I've added a commit fixing this.

christinehc committed:
fix,format: `pd.np` -> `np` correction. apply linting/formatting

christinehc commented on snekmer/rules/learn.smk:
```diff
@@ -137,1224 +154,185 @@ if config["learnapp"]["fragmentation"]:
-        data=join("output", "{prefix}", "{nb}.npz"),
-        kmerobj=join("output", "kmerize_{prefix}", "{nb}.kmers"),
+        data=join(out_dir, "{prefix}", "{nb}.npz"),
+        kmerobj=join(out_dir, "kmerize_{prefix}", "{nb}.kmers"),
```
Why are these stored in "kmerize_vector" instead of just "kmerize"? For organizational purposes, please change this to "kmerize/vector" and "kmerize/vector_frag" instead of creating separate top-level directories

christinehc commented on snekmer/rules/learn.smk:
```diff
@@ -137,1224 +154,185 @@ if config["learnapp"]["fragmentation"]:
+        kmerobj=join(out_dir, "kmerize_{prefix}", "{nb}.kmers"),
     log:
-        join("output", "{prefix}_kmerize", "log", "{nb}.log"),
+        join(out_dir, "{prefix}_kmerize", "log", "{nb}.log"),
```
I'd like this updated to match the above comment for subdirectories rather than separate top-level directories, but this is also a bug and creates an empty "{prefix}_kmerize" directory with just the logs (should have been "kmerize_{prefix}")

christinehc commented on snekmer/rules/learn.smk:
General comment (which echoes the suggestions in the other comments in this file): the learn workflow, especially with `fragmentation=True`, creates many directories in the output folder. I'm getting eval_apply_frag, eval_apply_reversed_frag, eval_conf, fragmented, kmerize_vector, kmerize_vector_frag, learn, vector, vector_frag, vector_frag_kmerize, and vector_kmerize. (Sidenote: the last folder is also created erroneously due to a mismatch in naming.) This is a TON of output to sort through for the user. I would consolidate some of these by saving to subdirectories, i.e. "eval/apply_frag" + "eval/apply_reversed_frag", etc. but I would also suggest including a README for each output directory to explain the contents of each to the user because it's a bit of information overwhelm.

jjacobson95 replied:
Subdirectory/directory consolidation has been implemented for the normal pipeline as well as when fragmentation is set to True.  READMEs are not yet added

christinehc replied:
Looks much improved. Last minor nitpick is that I prefer to minimize abbreviations, e.g. `eval_conf` -> `evaluate_confidence` / `eval_confidence` / `confidence`

christinehc commented on snekmer/templates/apply_template.html:
I would add 1-2 sentences per fileset to clarify/summarize the contents of each. To me, this HTML ends up being quite basic and not informative.

Will we also be creating an learn template?

jjacobson95 replied:
Just a heads up that I am working on this so @tnitka and I don't accidentally duplicate work -
Currently creating a Learn template and I'll update the Apply template while I'm at it as well.

Both will include full reports and all of the information that users should need.   I think that this can take the place of the ReadMEs requested above as well?

jjacobson95 replied:
I updated the Learn and Apply reports so they look like this. I think this should be more clear than ~8 separate ReadMEs in each directory.

---

**Updated Report:**
<img width="1017" height="956" alt="Screenshot 2025-07-29 at 8 28 48 PM" src="https://github.com/user-attachments/assets/cd49ce01-b327-4104-8d0a-3f042de682b8" />

---

**Original Report:**
<img width="395" height="614" alt="Screenshot 2025-07-29 at 8 30 14 PM" src="https://github.com/user-attachments/assets/d262818d-3f43-4d4c-8395-2e115e4342c9" />

jjacobson95 replied:
I added/updated reports for Learn and Apply. I think this can take the place of the requested ReadMEs. It'd be about 8 separate ReadMEs across all of the directories, I think this is better, but let me know if you'd like both.


---

**New Apply Report** Screenshot:  (not full report)
The bold text is in the first paragraph is dynamic.

---

<img width="1048" height="960" alt="Screenshot 2025-07-29 at 8 35 52 PM" src="https://github.com/user-attachments/assets/3d6d6e38-c8f6-42e7-8c10-908d705a83fa" />

---

**Old Apply Report** Screenshot:

---

<img width="402" height="591" alt="Screenshot 2025-07-29 at 8 35 28 PM" src="https://github.com/user-attachments/assets/b67c3a75-a66a-4945-9deb-4f563b818ca9" />

christinehc replied:
I love these! Not to add more work, but can we apply whatever formatting/framework/stylesheet to the other Snekmer HTML templates as well?

jjacobson95 replied:
Yes but maybe in a separate PR so this one can proceed?  It was HTML (format), CSS (style), Jinja (template), and a lil bit of Javascript (for the show more button)!

christinehc reviewed (changes requested):
I've suggested a few changes to improve general usability. Please let me know if you have questions. Most of these are organizational/documentation change suggestions.

tnitka committed:
move Learn/Apply tutorial files to resources/tutorial/learnapp_tutorial_files

tnitka committed:
remove learn outputs and optimize apply input data

jjacobson95 committed:
Changed Family to family across all files

jjacobson95 committed:
more Family to family changes

jjacobson95 committed:
Fixed bug causing extra output dirs such as vector_frag_kmerize and vector_kmerize

jjacobson95 committed:
snakemake has the worst error messages... but I got the directory stuff figured out

jjacobson95 committed:
snakefmt run

jjacobson95 committed:
Further directory organization

jjacobson95 committed:
changed learnapp to learn_apply in all config and files that call config params

jjacobson95 committed:
reapplied snakefmt

jjacobson95 committed:
trying to get snakefmt to work, make some dummy changes:

jjacobson95 committed:
manually making changes until snakefmt likes it

jjacobson95 committed:
Created full reports for Learn and Apply. Changed outputs slightly, from dashes in names to underscores.

jjacobson95 committed:
i always forget to run snakefmt

jjacobson95 committed:
manual linting done. my snakefmt does not work

jjacobson95 committed:
i hate snakefmt. manually made changes again

jjacobson95 committed:
mini changes to text in report

jjacobson95 committed:
Provided description of annotation file with matching fasta sequence id in tutorial

jjacobson95 committed:
Made requested changes to config files. Removed extras

jjacobson95 commented:
All requested changes have been made or addressed in comments. Largest changes were a major update on the directory structure generated in the output as well as comprehensive and more useful reports generated for both Learn and Apply.

Please let me know if there are any additional requested changes

christinehc commented on docs/source/tutorial/snekmer_learnapp_tutorial.ipynb:
```diff
@@ -36,22 +36,48 @@
-    "    conda install -c anaconda ipykernel \n",
-    "    python -m ipykernel install --user --name=snekmer\n",
-    "    jupyter notebook\n",
+    "    `conda activate snekmer`\n",
```
This isn't formatting correctly; I would use a codeblock

jjacobson95 replied:
Kept in markdown because its bash but reformatted so it is not all one line.

christinehc commented on snekmer/templates/apply_template.html:
See previous comment on applying this stylesheet to the other HTML templates :)

jjacobson95 replied:
Totally agree, and I'll add when I get the chance, but might be out of scope for this specific PR?

christinehc replied:
Got it, we can open a new issue/PR. Would like to bundle these changes into the next release, though :)

christinehc reviewed (approved):
Very minor changes suggested, but otherwise looks great! Thanks for the hard work :)

jjacobson95 committed:
change made to tutorial. Still markdown because bash but formatted correctly.

biodataganache commented:
Looks like we still have some more to go on this PR:
#133 
#134
#135
#136 
#137
(these are non-comprehensive: will be adding more)

jjacobson95 committed:
fixes for LA demo

jjacobson95 commented:
Just a note here, the conda environment just started failing in the CI for no apparent reason.  This has been (semi-)resolved
1) This was working in [e414c15](https://github.com/PNNL-CompBio/Snekmer/pull/127/commits/e414c1558882d2fd4dc88c223cdc2407867f283b)
2) This failed when I removed evaluate.smk. It said to add pephubclient.
3) I added the pephubclient. It failed even worse.
4) I force a revert to [e414c15](https://github.com/PNNL-CompBio/Snekmer/pull/127/commits/e414c1558882d2fd4dc88c223cdc2407867f283b), and it is passing again.

I'm not sure why removing evaluate.smk would cause a conda failure, but it seems like it does.

jjacobson95 committed:
updated pulp in requirements

jjacobson95 committed:
Attempting to move reporting code out of rules

jjacobson95 committed:
updates

jjacobson95 committed:
working on making code callable by demo

jjacobson95 committed:
working on making code callable externally (from demo)

jjacobson95 committed:
a bunch of updates for tutorial. Testing... if things break, revert this

jjacobson95 committed:
Okay I've made it so we can call most scripts from the new tutorial (which is in progress).

tnitka committed:
Pin scikit-learn and remove motif from Learn/Apply tutorial

tnitka commented:
This addresses #128 and #123.

jjacobson95 committed:
Okay. Major updates to the tutorial and code exports. A couple of code bug fixes that appeared with small datasets

jjacobson95 committed:
removed old test files

jjacobson95 committed:
added docker repo with info on how to build and run

jjacobson95 committed:
Updates to readme. small change to dockerfile

tnitka committed:
Update docs with venv installation instructions

tnitka committed:
Fix docs formatting

tnitka committed:
Actually fix formatting in docs

jjacobson95 committed:
Save point. Attempting to upgrade snakemake after here

tnitka committed:
Additional fixes to installation docs

tnitka committed:
Further fixes to install instructions

jjacobson95 committed:
Tutorial updates. Some variable name changes. And working on upgrading to snakemake9

jjacobson95 committed:
I think this fixes the cluster issue. search_report fix will be the next thing to work on

jjacobson95 committed:
Major file, demo, tutorial restructure and cleaning. templates updated. bugs fixed for snakemake version bump. These included scikit changes for newer python, recursive issues, casting, etc. Images moved to an image dir in resources

jjacobson95 committed:
small tutorial update

jjacobson95 committed:
New demo and inputs actually added

jjacobson95 committed:
more snakemake9 updates

jjacobson95 committed:
snakefmt applied. removed unsued code. Now need to test

jjacobson95 committed:
updated docs tutorial file

jjacobson95 committed:
Maybe the final update

jjacobson95 committed:
Making more windows friendly. Will have to make python scripts for demos too.

jjacobson95 committed:
Working for windows. demo changed to python because windows didn't work. Templates all updated for windows

jjacobson95 committed:
more updates

tnitka committed:
Update docs with new installation instructions

jjacobson95 committed:
update

jjacobson95 committed:
merge comflict resolution

jjacobson95 committed:
switch test back to learn_decoy_imp

jjacobson95 committed:
Our test uses an old linter version. Attempting to update the test

jjacobson95 committed:
the snakefmt version is unspecified and outdated in super-linter which is what our test uses. attempting to upgrade to their development branch because maybe its the new version there

jjacobson95 committed:
super-linter snakefmt is outdated and I turned it off.  Files have been linted.

jjacobson95 committed:
turned of lint step in test totally.

jjacobson95 committed:
Lint test properly off

jjacobson95 committed:
Trying to make test env compatible with python3.11 and new snakemake9+

jjacobson95 committed:
Environment looks like it built right. Updating snekmer tests now

jjacobson95 committed:
Environment dependancies are build correctly, but path not updated. Attempting

jjacobson95 commented:
**pre_paper_updates** branch has been merged
- Numerous updates to tutorials, demos, and the resource directory. 
- Code and dependancies now compatible with snakemake9.0+. **Now requires Python3.11+.**
- Several learn/apply bug fixes but no other changes that would affect learn/apply functionality.
- Report updates for clarify and compatibility
- Documentation updates
- Tested on mac, windows, ubuntu, linux. All use *requirements.txt* with the directions in the docs. (the environment.yml files could probably be removed)
- Githb tests updated to use venv and be compatible with the new additions
  - I turned off the super-lint step in the github tests because it does not have the latest version of snakefmt and causes more problems than it solves.  The .smk files have been linted on the command line.

tnitka committed:
Update install instructions with current Python version requirement (3.11+)

jjacobson95 committed:
Final update (i think). Just bumped the version in the dockerfile

jjacobson95 committed:
Merge remote-tracking branch 'refs/remotes/origin/learn_decoy_implementation' into learn_decoy_implementation

tnitka committed:
Remove outdated pip install instructions

tnitka committed:
Update conda environment files so conda still works

tnitka committed:
Fix error that prevented Learn/Apply tutorial from appearing in docs

tnitka committed:
docs: Update Learn/Apply tutorial to not use Conda

tnitka committed:
docs: update FAQ to reflect venv install recommendation

tnitka committed:
[skip ci] point action.yml back to main for merge
